From e80565d8402e8472617ed8528e558a3b9f0dfd08 Mon Sep 17 00:00:00 2001
From: Jianjian Huo <samuel.huo@gmail.com>
Date: Sun, 13 Jul 2014 09:08:59 -0700
Subject: [PATCH] bcache: add mutex lock for bch_is_open
References: bnc#902893
Commit-id: 789d21dbd9d8889e62c79ec19585fcc97e42ef07
Patch-Mainline: v3.17

Since bch_is_open will iterate linked list bch_cache_sets and
uncached_devices, it needs bch_register_lock.

Signed-off-by: Jianjian Huo <samuel.huo@gmail.com>
Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 drivers/md/bcache/super.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/md/bcache/super.c b/drivers/md/bcache/super.c
index f9d9045..574ace5 100644
--- a/drivers/md/bcache/super.c
+++ b/drivers/md/bcache/super.c
@@ -1893,10 +1893,12 @@ static ssize_t register_bcache(struct kobject *k, struct kobj_attribute *attr,
 	if (IS_ERR(bdev)) {
 		if (bdev == ERR_PTR(-EBUSY)) {
 			bdev = lookup_bdev(strim(path));
+			mutex_lock(&bch_register_lock);
 			if (!IS_ERR(bdev) && bch_is_open(bdev))
 				err = "device already registered";
 			else
 				err = "device busy";
+			mutex_unlock(&bch_register_lock);
 		}
 		goto err;
 	}
-- 
1.8.4.5

